---
title: "KnockoffZoom"
---

<style>
p.comment {
background-color: #DBDBDB;
padding: 10px;
border: 1px solid black;
margin-left: 25px;
border-radius: 5px;
}

</style>

KnockoffZoom is a flexible tool for the multi-resolution localization of causal variants across the genome.

## Overview

*KnockoffZoom* localizes causal variants precisely and provably controls the false discovery rate using artificial genotypes as negative controls. Our method is equally valid for quantitative and binary phenotypes, making no assumptions about their genetic architectures. Instead, we rely on well-established genetic models of linkage disequilibrium.

For more information, read the accompanying [paper](https://www.biorxiv.org/content/10.1101/631390v1):

<p class="comment">
<i>Multi-resolution localization of causal variants across the genome</i><br>
M. Sesia, E. Katsevich, S. Bates, E. Cand√®s, C. Sabatti<br>
bioRxiv 631390; doi: https://doi.org/10.1101/631390
</p>

&nbsp;

<p style="text-align:center;">
<img src="images/chicago.png" alt="KnockoffZoom discoveries"
	title="KnockoffZoom discoveries" width="80%" height="80%" />
</p>

We partition the genome into contiguous LD blocks and test whether the trait is independent of
the SNPs in any block, conditioning on the others. The resolution of the hypotheses is determined
by the size and homogeneity of the LD blocks, which we choose in advance.
To balance power and resolution, we consider multiple partitions, starting with a coarse view and successively  refining it. Some of the results obtained at different resolutions for the phenotype "height" in the UK Biobank are visualized below.
